### PR TITLE
Fix Watchdog disabling bug in WatchdogService.

### PR DESCRIPTION
**Pull request by pierrepitiot**
When the watchdog is disabled the thread is terminated before having disabled the watchdog. This PR fixes this bug.

### CODE REVIEW

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 1]
```diff
@@ -31,310 +31,310 @@
 
 public class WatchdogServiceImpl implements WatchdogService, ConfigurableComponent {
 
-    private static final Logger s_logger = LoggerFactory.getLogger(WatchdogServiceImpl.class);
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: I do think this PR doesn't use the new code style formatter.

Comment (pierrepitiot) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: I am very sorry but I never had any details on the new code style formatter and I have no idea of what I should do...

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 2]
```diff
@@ -35,130 +35,109 @@
...
-
-    protected void activate(ComponentContext componentContext, Map<String, Object> properties) {
-        this.m_properties = properties;
-        if (properties == null) {
+    private static ScheduledFuture<?>  		s_pollThreadTask;
+    private ScheduledExecutorService     	m_pollThreadExecutor;
+
+    private Map<String,Object>				m_properties;	
+
+    private int 							pingInterval = 2000;	//milliseconds
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: This still doesn't seem correct.

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 3]
```diff
@@ -35,130 +35,109 @@
...
+        m_properties=properties;
+        if(properties == null) {
             s_logger.debug("activating WatchdogService with null props");
         } else {
-            if (this.m_properties.get("enabled") != null) {
-                this.m_configEnabled = (Boolean) this.m_properties.get("enabled");
-                if (this.m_configEnabled) {
+            if(m_properties.get("enabled") != null){
+                m_configEnabled = (Boolean) m_properties.get("enabled");
+                if(m_configEnabled)
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: Same here

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 4]
```diff
@@ -35,130 +35,109 @@
...
                     s_logger.debug("activating WatchdogService with watchdog enabled");
-                } else {
+                else
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: and here

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 5]
```diff
@@ -233,25 +208,25 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
...
                 FileWriter bw = null;
                 try {
-                    f = new File(this.watchdogDevice);
+                    f = new File(watchdogDevice);
                     bw = new FileWriter(f);
                     bw.write('V');
-                    s_logger.info("watchdog stopped");
+                    s_logger.info("watchdog disabled");
                 } catch (IOException e) {
                     e.printStackTrace();
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: I would suggest to use proper logging here

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 6]
```diff
@@ -289,15 +249,14 @@ private void doWatchdogLoop() {
...
-                        e.printStackTrace();
+                        s_logger.error("IOException on refresh watchdog : {}", e.getMessage());
                     } finally {
                         try {
                             if (bw != null) {
                                 bw.close();
                             }
                         } catch (IOException e) {
-                            e.printStackTrace();
+                            s_logger.error("IOException on closing watchdog file : {}", e.getMessage());
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: Could you please show the full stack trace:

```java
s_logger.error("IOException on closing watchdog file", e);
```

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 7]
```diff
@@ -316,15 +275,39 @@ private void doWatchdogLoop() {
                     this.m_enabled = true;
                     s_logger.info("watchdog started");
                 } catch (IOException e) {
-                    e.printStackTrace();
+                    s_logger.error("IOException on start watchdog : {}", e.getMessage());
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: Same here, I overlooked this before. Sorry.

**Comment by ctron** [comment]
Any update on the synchronization issue?

File: kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java [hunk 8]
```diff
@@ -42,9 +42,9 @@
 
     private int pingInterval = 2000;	// milliseconds
     private static ArrayList<CriticalComponentImpl> s_criticalServiceList;
-    private boolean m_configEnabled = false;	// initialized in properties, if false -> no watchdog
-    private boolean m_enabled;
-    private boolean m_watchdogToStop = false;
+    volatile boolean m_configEnabled = false;	// initialized in properties, if false -> no watchdog
```
Comment (ctron) [inline kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java]: it should be "private volatile" then

**Comment by MMaiero** [comment]
This PR closes #684

**Comment by pierrepitiot** [comment]
thanks for the Java lesson Jens ;-)

**Pull request merged**